// abstract-sdk/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Client", {
  enumerable: true,
  get: function () {
    return _Client.default;
  }
});
Object.defineProperty(exports, "paginate", {
  enumerable: true,
  get: function () {
    return _paginate.default;
  }
});
Object.defineProperty(exports, "APITokenError", {
  enumerable: true,
  get: function () {
    return _errors.APITokenError;
  }
});
Object.defineProperty(exports, "BaseError", {
  enumerable: true,
  get: function () {
    return _errors.BaseError;
  }
});
Object.defineProperty(exports, "CLIPathError", {
  enumerable: true,
  get: function () {
    return _errors.CLIPathError;
  }
});
Object.defineProperty(exports, "EndpointUndefinedError", {
  enumerable: true,
  get: function () {
    return _errors.EndpointUndefinedError;
  }
});
Object.defineProperty(exports, "FileAPIError", {
  enumerable: true,
  get: function () {
    return _errors.FileAPIError;
  }
});
Object.defineProperty(exports, "ForbiddenError", {
  enumerable: true,
  get: function () {
    return _errors.ForbiddenError;
  }
});
Object.defineProperty(exports, "InternalServerError", {
  enumerable: true,
  get: function () {
    return _errors.InternalServerError;
  }
});
Object.defineProperty(exports, "NotFoundError", {
  enumerable: true,
  get: function () {
    return _errors.NotFoundError;
  }
});
Object.defineProperty(exports, "RateLimitError", {
  enumerable: true,
  get: function () {
    return _errors.RateLimitError;
  }
});
Object.defineProperty(exports, "ServiceUnavailableError", {
  enumerable: true,
  get: function () {
    return _errors.ServiceUnavailableError;
  }
});
Object.defineProperty(exports, "UnauthorizedError", {
  enumerable: true,
  get: function () {
    return _errors.UnauthorizedError;
  }
});
exports.sketch = void 0;

var _Client = _interopRequireDefault(require("./Client"));

var _paginate = _interopRequireDefault(require("./paginate"));

var sketch = _interopRequireWildcard(require("./sketch"));

exports.sketch = sketch;

var _errors = require("./errors");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
