// abstract-sdk/lib/endpoints/Activities.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _queryString = _interopRequireDefault(require("query-string"));

var _Cursor = _interopRequireDefault(require("../Cursor"));

var _Endpoint = _interopRequireDefault(require("./Endpoint"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Activities extends _Endpoint.default {
  info(descriptor) {
    return this.request({
      api: () => {
        return this.apiRequest(`activities/${descriptor.activityId}`);
      },
      cache: {
        key: `activity:${descriptor.activityId}`
      }
    });
  }

  list(descriptor, options = {}) {
    return this.request({
      api: () => {
        return new _Cursor.default(async (meta = {
          nextOffset: options.offset
        }) => {
          const query = _queryString.default.stringify(_objectSpread({}, descriptor, options, {
            offset: meta.nextOffset
          }));

          const response = await this.apiRequest(`activities?${query}`);
          return {
            data: response.data.activities,
            meta: response.meta
          };
        });
      }
    });
  }

}

exports.default = Activities;
