// abstract-sdk/lib/endpoints/Changesets.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Endpoint = _interopRequireDefault(require("./Endpoint"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Changesets extends _Endpoint.default {
  async info(descriptor) {
    const latestDescriptor = await this.client.descriptors.getLatestDescriptor(descriptor);
    return this.request({
      api: async () => {
        const response = await this.apiRequest(`projects/${latestDescriptor.projectId}/branches/${latestDescriptor.branchId}/commits/${latestDescriptor.sha}/changeset`);
        return response.changeset;
      },
      cli: () => {
        return this.cliRequest(["changeset", latestDescriptor.projectId, "--commit", latestDescriptor.sha, "--branch", latestDescriptor.branchId]);
      },
      cache: {
        key: `changeset:${descriptor.sha}`,
        disabled: descriptor.sha === "latest"
      }
    });
  }

}

exports.default = Changesets;
