// abstract-sdk/lib/endpoints/Data.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Endpoint = _interopRequireDefault(require("./Endpoint"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Data extends _Endpoint.default {
  async info(descriptor) {
    const latestDescriptor = await this.client.descriptors.getLatestDescriptor(descriptor);
    return this.request({
      api: () => {
        return this.apiRequest(`projects/${latestDescriptor.projectId}/branches/${latestDescriptor.branchId}/commits/${latestDescriptor.sha}/files/${latestDescriptor.fileId}/layers/${latestDescriptor.layerId}/data`);
      },
      cli: () => {
        return this.cliRequest(["layer", "data", latestDescriptor.projectId, latestDescriptor.sha, latestDescriptor.fileId, latestDescriptor.layerId]);
      },
      cache: {
        key: `layer-data:${descriptor.layerId}`,
        disable: descriptor.sha === "latest"
      }
    });
  }

}

exports.default = Data;
