// abstract-sdk/lib/endpoints/Files.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _errors = require("../errors");

var _utils = require("../utils");

var _Endpoint = _interopRequireDefault(require("./Endpoint"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

class Files extends _Endpoint.default {
  async info(descriptor) {
    const latestDescriptor = await this.client.descriptors.getLatestDescriptor(descriptor);
    return this.request({
      api: async () => {
        const {
          fileId
        } = latestDescriptor,
              branchDescriptor = _objectWithoutProperties(latestDescriptor, ["fileId"]);

        const files = await this.list(branchDescriptor);
        const file = files.find(file => file.id === latestDescriptor.fileId);

        if (!file) {
          throw new _errors.NotFoundError(`pageId=${latestDescriptor.fileId}`);
        }

        return file;
      },
      cli: async () => {
        const response = await this.cliRequest(["file", latestDescriptor.projectId, latestDescriptor.sha, latestDescriptor.fileId]);
        return response.file;
      },
      cache: {
        key: `file:${descriptor.fileId}`,
        disable: descriptor.sha === "latest"
      }
    });
  }

  async list(descriptor) {
    const latestDescriptor = await this.client.descriptors.getLatestDescriptor(descriptor);
    return this.request({
      api: async () => {
        const response = await this.apiRequest(`projects/${latestDescriptor.projectId}/branches/${latestDescriptor.branchId}/files`);
        return response.files;
      },
      cli: async () => {
        const response = await this.cliRequest(["files", latestDescriptor.projectId, latestDescriptor.sha]);
        return response.files;
      }
    });
  }

  async raw(descriptor, options = {}) {
    const latestDescriptor = await this.client.descriptors.getLatestDescriptor(descriptor);
    return this.request({
      cli: async () => {
        /* istanbul ignore if */
        if (!(0, _utils.isNodeEnvironment)() || options.disableWrite) {
          return;
        }

        this.cliRequest(["file", "export", latestDescriptor.fileId, options.filename || process.cwd(), `--project-id=${latestDescriptor.projectId}`, `--branch-id=${latestDescriptor.branchId}`, `--sha=${latestDescriptor.sha}`]);
      }
    });
  }

}

exports.default = Files;
