// abstract-sdk/lib/endpoints/Shares.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _utils = require("../utils");

var _Endpoint = _interopRequireDefault(require("./Endpoint"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const headers = {
  "Abstract-Api-Version": "13"
};

class Activities extends _Endpoint.default {
  create(descriptor, shareInput) {
    return this.request({
      api: async () => {
        return this.apiRequest("share_links", {
          method: "POST",
          body: _objectSpread({}, descriptor, shareInput, {
            commitSha: shareInput.sha
          }),
          headers
        });
      }
    });
  }

  info(descriptor) {
    return this.request({
      api: () => {
        return this.apiRequest(`share_links/${(0, _utils.inferShareId)(descriptor)}`, {
          headers
        });
      },
      cache: {
        key: `share:${(0, _utils.inferShareId)(descriptor)}`
      }
    });
  }

}

exports.default = Activities;
